File: src/com/xp/solutions/_3_Longest_Substring_Without_Repeating_Characters/Solution.java
package com.xp.solutions._3_Longest_Substring_Without_Repeating_Characters;

/**
 * Description:
 * User: yukong
 * Date: 2018-01-15
 * Time: 15:14
 */
public class Solution {
    public  int lengthOfLongestSubstring(String s) {
        int idx = 0;  // 遍历下标
        int maxLength = 0; // 最大长度
        int maxStart = 0;  // 最大长度的开始下标
        int length = 0; // 遍历时无重复字符串的长度
        int start = 0;  //遍历时无重复字符串的开始下标
        char[] hash = new char[256]; //字符是否重复的hash表 默认0未访问  1 已访问

        while (idx < s.length()) {
            if ( hash[s.charAt(idx)] == 1) {
                if (length > maxLength) {
                    maxLength = length;
                    maxStart = start;
                }

                //重置之前已访问的标记
                while (s.charAt(start) != s.charAt(idx)) {
                    hash[s.charAt(start)] = 0;  //重置
                    length--;
                    start++;
                }
                start++;

            } else{
                length++;
                hash[s.charAt(idx)] = 1;  //标记为已访问
            }
            idx++;
        }

        if (length > maxLength) {
            maxLength = length;
            maxStart = start;
        }
        return maxLength;
    }


}

File: src/com/xp/solutions/_3_Longest_Substring_Without_Repeating_Characters/Solution.kt
package com.xp.solutions._3_Longest_Substring_Without_Repeating_Characters

import kotlin.math.max

/**
 *
 * 滑动窗口问题
 * 分为双指针 left right
 * right指针向前移动保证当前子串是不重复的
 * 如 遇到重复字符串(通过hashmap判断当前字符是否重复出现)
 * 则left向前移动直至 当前子串没有重复字符(right -left)
 * 记录途中的(right -left) 其中最大的就是最长不重复子串长度
 * @author yukong
 * @date 2020/3/8 20:12
 */
fun main() {
    println(lengthOfLongestSubstring("pwwkew"))
}

fun lengthOfLongestSubstring(s: String): Int {
    var left = 0
    var right = 0
    val indexMap = mutableMapOf<Char, Int>()
    var maxLength = 0
    while (left < s.length && right < s.length) {
        val ch = s[right]
        var count = indexMap.getOrDefault(ch, 0)
        indexMap[ch] = ++count
        right++
        while (indexMap[ch]!! > 1) {
            val newCh = s[left]
            count = indexMap.getOrDefault(newCh, 0)
            indexMap[s[left++]] = --count
        }
        maxLength = max(maxLength, right - left)
    }
    return maxLength
}